ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -733,52 +741,83 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
     long totalCapacity = tgtLog.getCapacityInBytes();
     long writtenLastTime = 0;
     try (FileChannel fileChannel = Utils.openChannel(logSegmentToCopy.getView().getFirst(), false)) {
-      for (IndexEntry srcIndexEntry : srcIndexEntries) {
-        IndexValue srcValue = srcIndexEntry.getValue();
-        long usedCapacity = tgtIndex.getLogUsedCapacity();
-        if (isActive && (tgtLog.getCapacityInBytes() - usedCapacity >= srcValue.getSize())) {
-          fileChannel.position(srcValue.getOffset().getOffset());
-          Offset endOffsetOfLastMessage = tgtLog.getEndOffset();
-          // call into diskIOScheduler to make sure we can proceed (assuming it won't be 0).
-          diskIOScheduler.getSlice(DiskManager.CLEANUP_OPS_JOB_NAME, DiskManager.CLEANUP_OPS_JOB_NAME, writtenLastTime);
-          tgtLog.appendFrom(fileChannel, srcValue.getSize());
-          FileSpan fileSpan = tgtLog.getFileSpanForMessage(endOffsetOfLastMessage, srcValue.getSize());
-          if (srcValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
-            IndexValue putValue = tgtIndex.findKey(srcIndexEntry.getKey());
-            if (putValue != null) {
-              tgtIndex.markAsDeleted(srcIndexEntry.getKey(), fileSpan, srcValue.getOperationTimeInMs());
+      ListIterator<IndexEntry> srcIndexEntryIterator = srcIndexEntries.listIterator();
+      ByteBuffer bufferToUse = bundleReadBuffer;
+      while (srcIndexEntryIterator.hasNext()) {
+        // reset startOffset to begin a new bundle read.
+        long startOffset = -1;
+        List<IndexEntry> readyEntries = new ArrayList<>();
+        // get entries as many as possible and do a bunch read.
+        while (srcIndexEntryIterator.hasNext()) {
+          IndexEntry currentEntry = srcIndexEntryIterator.next();
+          if (startOffset == -1) {
+            startOffset = currentEntry.getValue().getOffset().getOffset();
+            readyEntries.clear();
```

Comment:
is this required? Won't it be a new array list already?

Reply:
yes.